Unified Agents Blog Post (#9012)
Microsoft/vscode-docs: blogs/2025/11/03/universe-recap.md

> # A unified experience for all coding agents

Reviewer: Reminder to include frontmatter

If we had to pick one word to describe the past year, it would probably be "Agent".

> Agents took over VS Code in 2025. We released the [Agent Mode](https://code.visualstudio.com/blogs/2025/02/24/introducing-copilot-agent-mode) for VS Code, integration for the [Copilot Coding Agent (cloud)](https://code.visualstudio.com/docs/copilot/copilot-coding-agent), and the new [GitHub Copilot CLI](https://github.com/features/copilot/cli/). But Copilot is not the only agent game in town. There are now more coding agents than ever - including options from OpenAI and Anthropic.

Reviewer: ```suggestion
Agents took over VS Code in 2025. We released [agent mode](https://code.visualstudio.com/blogs/2025/02/24/introducing-copilot-agent-mode) for VS Code, integration for the [Copilot coding agent (cloud)](https://code.visualstudio.com/docs/copilot/copilot-coding-agent), and the new [GitHub Copilot CLI](https://github.com/features/copilot/cli/). But Copilot is not the only agent game in town. There are now more coding agents than ever - including options from OpenAI and Anthropic.
```

> With all these choices, things got better - and a little more fragmented. Subscription hopping, tool juggling, and FOMO became the norm. This year at GitHub Universe, we set out to fix that with a unified agent experience in VS Code. Now you can switch between agents - local or cloud - and manage them all from one place.

Reviewer 2: based on this paragraph, it feels like the natural segue here would be to talk about agent sessions in the next section. if keep open ai codex as the next section (which i'm a fan of as that's the biggest headline news imo) then i'd maybe add an additional mention at the end of how it's not just copilot agents we can see here, to make it a smoother transition

With all these choices, things got better for developers. But the agent ecosystem got a little more fragmented. Subscription hopping, tool juggling, and the constant FOMO on the latest agent trend is now the norm. This year at GitHub Universe, we set out to fix that with a unified agent experience in VS Code. The first big step towards making that a reality was offering more agents in your Copilot subscription. And not just those with "Copilot" in their name.

### OpenAI Codex Integration

> OpenAI had a big year: they shipped the GPT-5 and GPT-5 Codex models, which were available in VS Code on day one through the standard model picker. But they also launched Codex - OpenAI's coding agent - available as both A CLI tool and a [VS Code extension](https://marketplace.visualstudio.com/items?itemName=openai.chatgpt). And it was a huge hit with developers.

Reviewer: ```suggestion
OpenAI had a big year: they shipped the GPT-5 and GPT-5 Codex models, which were available in VS Code on day one through the standard model picker. But they also launched Codex, their coding agent, available as both a CLI tool and a [VS Code extension](https://marketplace.visualstudio.com/items?itemName=openai.chatgpt). And it was a huge hit with developers.
```

> This year at GitHub Universe, we announced you can now use OpenAI Codex with your GitHub Copilot Pro+ subscription. No additional subscription required.

Reviewer: Removing repeat language.

```suggestion
At GitHub Universe, we announced you can now use OpenAI Codex with your GitHub Copilot Pro+ subscription. No additional subscription required.
```

![OpenAI Codex sign-in panel in VS Code](codex-signin.gif)

When you use Codex with Copilot Pro+, Copilot handles all model calls and standard rate limits apply. You get code generation, code explanation, and all the features - no need to manage a separate OpenAI account.

With Codex, you now have four powerful coding agents in VS Code:

1. GitHub Copilot

> 2. GitHub Copilot Coding Agent (Cloud)

Reviewer 3: I think now we just call it "Cloud Agent"

> 1. GitHub Copilot

Reviewer: Probably no need to use a numbered list here

That's why we've introduced a new feature in VS Code for orchestrating all your agents - local or remote. We call it, "Agent Sessions".

### Agent Sessions

There's a new view in the VS Code sidebar called "[Agent Sessions](https://code.visualstudio.com/updates/v1_105#_agent-sessions)". It gives you one place to manage all your agents - local or cloud.

![VS Code window with Agent Sessions sidebar showing Copilot, Coding Agent, CLI, and Codex statuses against a calm gray workspace](agent-sessions.png)

With Agent Sessions, you see all agent sessions for your project. You can check which agents are running, their status, and jump between sessions with a click.

Cloud Agent sessions now have a tabbed experience that lets you course-correct the agent mid-run. It's common to send a prompt and realize you forgot something important. Before, you had to wait or cancel. Now, just open the tab, add an update, and watch the agent adjust its plan.

You can also delegate any task to any agent right from the chat.

![VS Code showing the "Delegate" button from the chat, when clicked opens a menu of agents to delegate to](why-settle.gif)

> This unified Agent Sessions view makes VS Code a "mission control" for orchestrating all your agents, while keeping you in the editor where you do your best work.

Reviewer 3: We should also mention that more agents are on the way.. i think there was a screenshot from GitHub Universe... maybe pibogg can telll us what we can say.

This unified Agent Sessions view makes VS Code a "mission control" for orchestrating all your agents, while keeping you in the editor where you do your best work. We're excited to welcome OpenAI Codex today, and we're working to bring more agents to your Copilot+ subscription in the future.

### Planning Agent

> A few months ago we introduced the concept of Chat Modes in VS Code. These are custom modes that let you augment or alter the behaviour of the built-in agent prompt. When you use a Chat Mode to alter the agent behaviour in VS Code, what you're _really_ doing is creating your own custom agent. So we've renamed "Chat Modes" to just "Agents" to better reflect what they actually are.

Reviewer: ```suggestion
A few months ago we introduced the concept of chat modes in VS Code. These are custom modes that let you augment or alter the behavior of the built-in agent prompt. When you use a chat mode to alter the agent behavior in VS Code, what you're _really_ doing is creating your own custom agent. So we've renamed "chat modes" to just "agents" to better reflect what they actually are.
```

![Copilot chat in VS Code with Plan agent dropdown highlighted, planning guidance beside dark theme editor, label reads Plan for a focused tone.](planning-agent.png)

The new Plan agent helps create a detailed plan from lazy prompts like "add drag and drop". That's an actual prompt I sent yesterday. No mention of what to add it to, what page, or whether to use a library. I do this a lot, and I bet I'm not alone.

With the Plan agent, Copilot asks the questions that need answers. It even recommends libraries for drag and drop and gives reasons to pick one over another.

![Plan agent breaking down drag-and-drop into steps recommending React Beautiful DnD and React DnD with comparisons.](plan-in-action.png)

You can answer these with quick replies on separate lines so it knows which answer goes to which question. Here's how I'd answer:

```
dnd-kit
yes - what kind of a question is this in 2025
link creation only
```

> Pro tip: Change the "workbench.action.chat.submit" keybinding to "Ctrl + Enter" so you stop accidentally sending messages when you just want a new line. Your swear jar will thank you.

> When the Plan agent has enough info, it stops asking questions and asks if you're ready to proceed. You can use the new "[Handoff](https://code.visualstudio.com/updates/v1_105#_handoffs)" feature in chat to either proceed or open the full plan in the editor.

Reviewer: ```suggestion
When the Plan agent has enough info, it stops asking questions and asks if you're ready to proceed. You can use the new "[Handoff](https://code.visualstudio.com/docs/copilot/customization/custom-chat-modes#_handoffs)" feature in chat to either proceed or open the full plan in the editor.
```

![Screenshot showing the Handoff feature in Copilot chat with options to proceed with implementation or open the plan in the editor.](handoffs.png)

Try different models to see which you like best for planning. We've found the [Claude models](https://www.anthropic.com/claude) are great at identifying missing context and edge cases, and asking the right questions.

If you're like me, you'll want to know how the Plan agent works so you can up your prompt engineering game. You can read the Plan prompt by choosing "Configure Agents" from the Command Palette and selecting Plan. It's a great baseline for creating your own [custom agents](https://code.visualstudio.com/updates/v1_105#_support-for-nested-agentsmd-files-experimental). I used it to create one called ["Research"](https://gist.github.com/burkeholland/919d655ae4df5c809b549632c3afb144) that recursively does internet research and writes up its findings.

> > Pro tip: You can find hundreds of custom instructions, prompt files and agents over on the [awesome-copilot](https://github.com/github/awesome-copilot) repo. If you haven't checked that out yet, you're missing out. It's a treasure trove of inspiration and ready-made prompts.

Reviewer 3: An additional cool thing is that you can use these same agents in the CLI and with Cloud Agent as well when you assign or delegate a task on github.com. Better together

> These custom agents are also available when you delete to other agents such as the CLI and the Cloud. Your custom agents work everywhere that you need them to.

Reviewer: ```suggestion
These custom agents are also available when you delegate to other agents such as the Copilot CLI and the Copilot coding agent. Your custom agents work everywhere that you need them to.
```

When a subagent finishes, it returns the final result to the main chat - and only that result joins the main context. Subagents keep your main chat lean while letting you go on sidebars and deep dives. For example, if you're building an API and need to research authentication, spin up a subagent to do that.

```
Analyze the #file:api with #runSubagent and recommend the best authentication strategy for a web client consuming these endpoints.
```

You'll know a subagent is running because you can see tool calls and model responses below the subagent action. In the screenshot below, that's "Analyze app structure for auth".

![A subagent process running in VS Code with tool calls underneath the main agent action](subagent-in-progress.png)

We're still exploring ways to help you manage context with agents, and subagents are just the beginning.

## Looking Ahead

Agents are changing how we write code and how we work. You shouldn't have to pick just one. You should be able to move between agents, keep fine-grained control over your context, and create your own Agents to extend our system prompts.

> These are just a few highlights from this year's [GitHub Universe](https://github.com/events/universe). Check out [GitHub's blog](https://github.blog/) for all the updates as we work on a unified workflow for a multi-agent experience everywhere you need it.

Reviewer 2: is github's blog the main CTA we want to have here? feels like we should be linking to vs code release notes or blog instead

> These are just a few highlights from this year's [GitHub Universe](https://github.com/events/universe). Check out [GitHub's blog](https://github.blog/) for all the updates as we work on a unified workflow for a multi-agent experience everywhere you need it.

Reviewer 3: maybe also the changelog?

> Agents are changing how we write code and how we work. You shouldn't have to pick just one. You should be able to move between agents, keep fine-grained control over your context, and create your own custom Agents to extend the various built-in agent prompts. With the unified agent experience in VS Code, you can now do all of that.

Reviewer: ```suggestion
Agents are changing how we write code and how we work. You shouldn't have to pick just one. You should be able to move between agents, keep fine-grained control over your context, and create your own custom agents to extend the various built-in agent prompts. With the unified agent experience in VS Code, you can now do all of that.
```

---
Order: 108
TOCTitle: A Unfied Agent Experience
PageTitle: A Unified Experience for all Coding Agents
MetaDescription: "Agents took over VS Code in 2025. We released agent mode for VS Code, integration for the Copilot coding agent, and the new GitHub Copilot CLI. But Copilot is not the only agent game in town. There are now more coding agents than ever, including options from OpenAI and Anthropic."
MetaSocialImage: expanding-model-choice.png
Date: 2025-11-04
Author: Burke Holland
---

> 

Reviewer: ```suggestion
# A Unified Experience for all Coding Agents

November 4, 2025 by Burke Holland, [@burkeholland](https://twitter.com/burkeholland)

```

# Agent sessions, A planning agent and OpenAI Codex Integration: A Universe Recap

> VS Code releases 11 times a year. That's a release every _single_ month.  And all of those releases culminate each year with our announcements at GitHub Universe. Last year, we announced that Claude was being added to Copilot and a new chat mode called "Edits" that would both answer your chat questions AND write code. What a time to be alive!

Reviewer 3: may clarify that it is the model being added? else confusion around claude code now

![OpenAI Codex sign-in panel in VS Code](codex-signin.gif)

When you use Codex via your Copilot Pro+ account, Copilot is used for all the model calls and the standard Copilot rate limits and billing apply. This means you can use Codex for code generation, code explanation, and all the other great features it provides without needing to manage a separate OpenAI account or subscription.

This is a huge step forward in unifying the agent ecosystem. Developers can now choose between multiple top-tier coding agents - Copilot and Codex - all within the same subscription and experience.

With the addition of Codex, you now have access to four incredibly capable coding agents within VS Code:

1. GitHub Copilot
2. GitHub Copilot Coding Agent (Cloud)
3. GitHub Copilot CLI
4. OpenAI Codex

But with all of these agents available, it can be easy to get overwhelemed and lost in the sauce. What agents are running again? Where are they running? What day is it?

And that's why we've introduced a new feature in VS Code for orchestrating all your agents - local or remote. And we call it, Agent Sessions.

> ### Agent Sessions

Author: I don't think there is?

```
dnd-kit
yes - what kind of a question is this in 2025
link creation only
```

> Pro tip: Change the "workbench.action.chat.submit" keybinding to "Ctrl + Enter" so that you can stop accidentally sending messages when you just want a new line. Your swear jar thanks you.

When the Plan agent feels like it has enough information, it will stop asking questions and ask you if you are ready to proceed with the implementation. You can use the new "Handoff" feature in chat to either proceed with the implementation or open the full plan in the editor.

![Screenshot showing the Handoff feature in Copilot chat with options to proceed with implementation or open the plan in the editor.](handoffs.png)

You can experiment with different models to see which one you like beset for planning. We've found the Claude models to be exceptionally good at planning out tasks because they are especially skilled in at proactively identifying missing context and edge cases and asking the right questions to fill in those gaps.

> If you're like me, you're going to want to know how this Plan agent works so that you can up your prompt engineering game. You can examine the Plan prompt by choosing "Configure Agents" from the Command Palette and selecting the Plan prompt. I recommend reading through this prompt so that you can see how we define the looping workflow, use XML tags to direct the model, provide examples, etc. It also makes a great baseline for creating your own agents. I used it to create one called ["Research"](https://gist.github.com/burkeholland/919d655ae4df5c809b549632c3afb144) that recursively does internet research and writes up its findings.

Reviewer 3: we sshould point ot awesome copilot as we launched partner agents already and you can use then on github as well. https://montemagno.com/building-better-apps-with-github-copilot-custom-agents/
